man/yokadi.1
.TH YOKADI 1 "July 10, 2009"
.SH NAME
yokadi \- commandline todo system
.SH SYNOPSIS
.B yokadi
.RI [ options ]...
.br
.SH DESCRIPTION
.B yokadi
is a command-line oriented, SQLite powered, TODO list tool.  It helps
you organize all the things you have to do and you must not forget. It aims to
be simple, intuitive and very efficient. In Yokadi you manage projects, which
contains tasks. At the minimum, a task has a title, but it can also have a
description, a due date, an urgency or keywords. Keywords can be any word that
help you to find and sort your tasks.
.PP
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-datadir=<DIR>
Database directory.
.TP
.B \-c, \-\-create-only
Just create an empty database.
.TP
.B \-u, \-\-update
Update database to the latest version.
.TP
.B \-h, \-\-help
Show summary of options and exit.
.TP
.B \-v, \-\-version
Show version of program and exit.
.SH SEE ALSO
.BR yokadid (1).
.br
.SH SEE ALSO
Website: http://yokadi.github.io

Mailing List: http://sequanux.org/cgi-bin/mailman/listinfo/ml-yokadi
.SH AUTHOR
yokadi was written by Aurélien Gâteau <mail@agateau.com> and Sébastien Renard <sebastien.renard@digitalfox.org>.
.PP
This manual page was written by Kartik Mistry <kartik@debian.org>,
for the Debian project (and may be used by others).


NEWS
v1.2.0 unreleased
- New features:
    - The new `p_merge` command lets you merge a project into another.
    - It is now possible to turn a task into a note with `t_to_note` and a note into a task with `n_to_task`.
- Bug fixes:
    - The `k_remove` command no longer ignores unused keywords.
    - HTML output has been fixed to no longer output strings wrapped in `b""`.
    - `t_list` filtering has been fixed so that `t_list --urgency 0` filters out tasks with a negative urgency, as expected.
- Improvements:
    - HTML output has been refreshed:
        - It looks more modern now.
        - Some fields have been removed (doneDate, creationDate).
        - The title, keywords and description fields have been merged.
        - An ID field has been added (handy to run a command on a task listed in the output).
        - Columns now use human-friendly titles.
- Misc:
    - The `--db` option is now deprecated and replaced by the `--datadir` option. `--db` will be removed in the next version.
    - Similarly, the `YOKADI_DB` environment variable is now deprecated and will be removed in the next version.
    - Yokadi no longer supports cryptography: encrypted databases will be decrypted at update.

v1.1.1 2016/11/11

- Improvements:
    - When listing multiple projects, order them alphabetically.
- Bug fixes:
    - Fixed parse error if the user sets a time of "17m".
    - When the user edits a tasks with t_edit and removes a keyword, remove the keyword from the task.
    - Made recurrence code work with dateutil 2.6.0.

v1.1.0 2016/09/03

- New features & Improvements:
    - A new command has been added: `t_medit`. `t_medit` lets you edit all tasks of a project in one go.
    - Aliases can now be modified. The name of the alias can be modified with `a_edit_name` and the command with `a_edit_command`.
    - Database format updates are now easier to run: just run `yokadi -u`, no more separate `update.py` command. Updates are also much faster.
    - Task lists have been improved:
        - Borders look nicer.
        - Some bugs in the rendering of the title column have been fixed (wrong width, badly cropped text).
    - Yokadi now uses standard paths by default: the database is stored in ~/.local/share/yokadi/yokadi.db and non-essential data is in ~/.cache/yokadi/.
    - Reviewed and improved documentation. Moved developer documentation to a separate dir (doc/dev).
- Bug fixes:
    - The code handling recurrences has been made more robust.
    - Recurrences are now stored in a more future proof way.
    - Fixed `bug_edit` crash.
    - Fixed negative keyword filter: A task with two keywords k1 and k2 would not be excluded by a filter !k1.

v1.0.2 2016/03/28

- Use a more portable way to get the terminal size. This makes it possible to use Yokadi inside Android terminal emulators like Termux
- Sometimes the task lock used to prevent editing the same task description from multiple Yokadi instances were not correctly released
- Deleting a keyword from the database caused a crash when a t_list returned tasks which previously contained this keyword

v1.0.1 2015/12/03

- User changes:
    - Make sure installing via pip installs the required dependencies

- Developer changes:
    - Improved release process

v1.0.0 2015/11/29

- User changes:
    - Fixed an issue which caused t_list to fail when filtering by keywords on large lists
    - Removed the project keywords feature. It was not very useful and made the searching code more complicated
    - Fixed ical support: it now works with ical 3.6 or later
    - Improved documentation
    - Added Keywords field to yokadi.desktop
- Developer changes:
    - Yokadi has been ported to Python 3
    - The application now uses SQLAlchemy instead of SQLObject to access the SQLite database

v0.14 2014/05/03
- Command changes:
    - t_add, n_add:
        - Allow creating two tasks with the same title (useful for recurrent tasks, like "buy bread").
        - Allow using _ to select last project, making it possible to do multiple t_add on the same project with `t_add _ <task description>`.
        - Add --describe option to start describing the task right after adding it.
    - t_describe, n_describe:
        - Safer task description editing: task is updated each time the editor saves, a lock manager now prevents multiple edits.
        - Use .md suffix instead of .txt for the temporary filename to allow some smart things with editors that understand Markdown.
        - Use project and task name for the temporary filename. Useful when using graphical editors or when your terminal title shows the current running command.
    - t_due:
        - When called with a time argument which is before current time, set due date to the day after.
    - t_show:
        - Show the task ID.
    - t_list:
        - Use month and year for the task age if the task is older than 12 months.
        - Add support for arbitrary minimum date for --done.
        - Fixed broken help.
    - n_list:
        - Display creation date instead of age.
        - Notes are now grouped by date.
    - p_list:
        - Show task count per project.
    - p_remove:
        - Show the number of associated tasks in the prompt.
    - p_edit:
        - Handle case where user tries to rename a project using the name of an existing project.

- yokadid:
    - Add --restart option and --log option.
    - Set process name with setproctitle.
    - Configuration keys can now be overridden using environment variables.

- Misc:
    - Date/time commands now support %d/%m/%y date format.
    - Replaced xyokadi with a desktop file.
    - Updated README to match real output.

- Developer specific changes:
    - Command parser has been ported from optparse to argparse.
    - Code is now PEP 8 compliant, with the exception of camelCase usage.
    - All imports have been changed to absolute imports (ie `import yokadi.<something>`).
    - Code has been reorganized into different sub directories.

src/yokadi/yokadi.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
"""
Command line oriented, sqlite powered, todo list

@author: Aurélien Gâteau <aurelien.gateau@free.fr>
@author: Sébastien Renard <Sebastien.Renard@digitalfox.org>
@license: GPL v3 or later
"""

import locale
import os
import sys
try:
    import readline
except ImportError:
    print "You don't have a working readline library."
    print "Windows users must install Pyreadline."
    print "Get it on https://launchpad.net/pyreadline/+download"
    print "Or use 'easy_install pyreadline'"
    sys.exit(1)

readline.parse_and_bind("set show-all-if-ambiguous on")
reload(sys)  ## So as to enable setdefaultencoding

import traceback
from cmd import Cmd
from optparse import OptionParser

try:
    from sqlobject import __doc__ as sqlobjectVersion

except ImportError:
    print "You must install SQLObject to use Yokadi"
    print "Get it on http://www.sqlobject.org/"
    print "Or use 'easy_install sqlobject'"
    sys.exit(1)

from core import db
from  ycli import tui
from core import utils
from core import cryptutils

from ycli.aliascmd import AliasCmd, resolveAlias
from ycli.confcmd import ConfCmd
from ycli.keywordcmd import KeywordCmd
from ycli.projectcmd import ProjectCmd
from ycli.taskcmd import TaskCmd
from core.yokadiexception import YokadiException, BadUsageException
from core.yokadioptionparser import YokadiOptionParserNormalExitException

# Force default encoding to prefered encoding
# This is mandatory when yokadi output is piped in another command
sys.setdefaultencoding(tui.ENCODING)


#TODO: move YokadiCmd to a separate module in ycli package
class YokadiCmd(TaskCmd, ProjectCmd, KeywordCmd, ConfCmd, AliasCmd, Cmd):
    def __init__(self):
        Cmd.__init__(self)
        TaskCmd.__init__(self)
        ProjectCmd.__init__(self)
        KeywordCmd.__init__(self)
        AliasCmd.__init__(self)
        ConfCmd.__init__(self)
        self.prompt = "yokadi> "
        self.historyPath = os.getenv("YOKADI_HISTORY")
        if not self.historyPath:
            if os.name == "posix":
                self.historyPath = os.path.join(os.path.expandvars("$HOME"), ".yokadi_history")
            else:
                # Windows location
                self.historyPath = os.path.join(os.path.expandvars("$APPDATA"), ".yokadi_history")
        self.loadHistory()
        self.cryptoMgr = cryptutils.YokadiCryptoManager() # Load shared cryptographic manager

    def emptyline(self):
        """Executed when input is empty. Reimplemented to do nothing."""
        return

    def default(self, line):
        nline = resolveAlias(line, self.aliases)
        if nline != line:
            return self.onecmd(nline)
        elif nline.isdigit():
            self.do_t_show(nline)
        elif nline == "_":
            self.do_t_show(nline)
        else:
            raise YokadiException("Unknown command. Use 'help' to see all available commands")

    def completedefault(self, text, line, begidx, endidx):
        """Default completion command.
        Try to see if command is an alias and find the
        appropriate complete function if it exists"""
        nline = resolveAlias(line, self.aliases)
        compfunc = getattr(self, 'complete_' + nline.split()[0])
        matches = compfunc(text, line, begidx, endidx)
        return matches

    def do_EOF(self, line):
        """Quit."""
        print
        return True

    #Some standard alias
    do_quit = do_EOF
    do_q = do_EOF
    do_exit = do_EOF

    def onecmd(self, line):
        """This method is subclassed just to be
        able to encapsulate it with a try/except bloc"""
        try:
            # Decode user input
            line = line.decode(tui.ENCODING)
            return Cmd.onecmd(self, line)
        except YokadiOptionParserNormalExitException:
            pass
        except UnicodeDecodeError, e:
            tui.error("Unicode decoding error. Please check you locale and terminal settings (%s)." % e)
        except UnicodeEncodeError, e:
            tui.error("Unicode encoding error. Please check you locale and terminal settings (%s)." % e)
        except BadUsageException, e:
            tui.error("*** Bad usage ***\n\t%s" % e)
            cmd = line.split(' ')[0]
            self.do_help(cmd)
        except YokadiException, e:
            tui.error("*** Yokadi error ***\n\t%s" % e)
        except IOError, e:
            # We can get I/O errors when yokadi is piped onto another shell commands
            # that breaks.
            print >> sys.stderr, "*** I/O error ***\n\t%s" % e
        except KeyboardInterrupt:
            print "*** Break ***"
        except Exception, e:
            tui.error("Unhandled exception (oups)\n\t%s" % e)
            print "This is a bug of Yokadi, sorry."
            print "Send the above message by email to Yokadi developers (ml-yokadi@sequanux.org) to help them make Yokadi better."
            cut = "---------------------8<----------------------------------------------"
            print cut
            traceback.print_exc()
            print "--"
            print "Python: %s" % sys.version.replace("\n", " ")
            print "SQLObject: %s" % sqlobjectVersion.replace("\n", " ")
            print "OS: %s (%s)" % os.uname()[0:3:2]
            print "Yokadi: %s" % utils.currentVersion()
            print cut
            print

    def loadHistory(self):
        """Tries to load previous history list from disk"""
        try:
            readline.read_history_file(self.historyPath)
        except Exception:
            # Cannot load any previous history. Start from a clear one
            pass

    def writeHistory(self):
        """Writes shell history to disk"""
        try:
            # Open r/w and close file to create one if needed
            historyFile = file(self.historyPath, "w")
            historyFile.close()
            readline.set_history_length(1000)
            readline.write_history_file(self.historyPath)
        except Exception, e:
            tui.warning("Fail to save history to %s. Error was:\n\t%s"
                        % (self.historyPath, e))

    def do_help(self, arg):
        """Type help <topic> to see the help for a given topic"""

        """
        Overload do_help to show help from the command parser if it exists:
        if there is a parser_foo() method, assume this method returns a
        YokadiOptionParser for the do_foo() method and show the help of the
        parser, instead of do_foo() docstring.
        """
        if arg in self.aliases:
            # If arg is an alias, get help on the underlying command
            arg = self.aliases[arg].split()[0]
        if hasattr(self, "parser_" + arg):
            parserMethod = getattr(self, "parser_" + arg)
            parserMethod().print_help(sys.stderr)
        else:
            print "Usage: ",
            Cmd.do_help(self, arg)

    def completenames(self, text, *ignored):
        """Complete commands names. Same as Cmd.cmd one but with support
        for command aliases. Code kindly borrowed to Pysql"""
        dotext = 'do_' + text
        names = [a[3:] for a in self.get_names() if a.startswith(dotext)]
        names.extend([a for a in self.aliases.keys() if a.startswith(text)])
        return names


def main():
    locale.setlocale(locale.LC_ALL, os.environ.get("LANG", "C"))
    parser = OptionParser()

    parser.add_option("-d", "--db", dest="filename",
                      help="TODO database", metavar="FILE")

    parser.add_option("-c", "--create-only",
                      dest="createOnly", default=False, action="store_true",
                      help="Just create an empty database")

    parser.add_option("-v", "--version",
                      dest="version", action="store_true",
                      help="Display Yokadi current version")

    (options, args) = parser.parse_args()

    if options.version:
        print "Yokadi - %s" % utils.currentVersion()
        return

    if not options.filename:
        # Look if user define an env VAR for yokadi db
        options.filename = os.getenv("YOKADI_DB")
        if options.filename:
            print "Using env defined database (%s)" % options.filename
        else:
            options.filename = os.path.normcase(os.path.expanduser("~/.yokadi.db"))
            print "Using default database (%s)" % options.filename

    db.connectDatabase(options.filename)

    if options.createOnly:
        return

    db.setDefaultConfig() # Set default config parameters

    cmd = YokadiCmd()
    try:
        if len(args) > 0:
            cmd.onecmd(" ".join(args))
        else:
            cmd.cmdloop()
    except KeyboardInterrupt:
        print "\n\tBreak ! (the nice way to quit is 'quit' or 'EOF' (ctrl-d)"
        sys.exit(1)
    # Save history
    cmd.writeHistory()

if __name__ == "__main__":
    main()
# vi: ts=4 sw=4 et
